### PR TITLE
[Update] Cool Stuff

### PR DESCRIPTION
**Pull request by ItsANoBrainer**
- Added support for saving the door dialog items instead of having to type them every time. This is a config option as well, but also requires a PR change to qb-input (https://github.com/qbcore-framework/qb-input/pull/17)
- Added support for using a custom door label instead of locked/unlocked text
- Added door label to qb-input dialog, renamed doorname to dooridentifier, and now the table key is saved as 'configname-identifier' instead of 'doorname'
- Added Configuration options to the config.lua
- Added cantUnlock option, which when set makes it so nobody can unlock a door (has to be done by the server), but anyone who passes authorization can lock it
- Added support for hideLabel option, hides the popup label for hiding doors
- Added hideLabel to ui options (this caused the input menu height to be too high, I included new changes in my other PR to address this)
- Added support for DoorDebugging, adds an optional config and command for showing 3DWorldText at the location of where the door 'center' is thought to be
- If label, pickable, cantunlock, or hidelabel are false then they are just removed from the doorData, to take up less config space, as its not needed anyway
- Fixed bug when trying to set a door to default UNLOCKED in which the string of 'false' was being used for the initial door state, which was coming as true and locking the door at creation
- Fixed fixText not working properly
- Added duplicate identifier check that stops you from creating a new door entry when a door with that identifier already exists in the system
- Volume code order fix
- Removed audioRemote from defaulting in the config (to false), waste of config space
- Remote trigger support (pressing H will raycast from camera and will try to get the closest door to the hit location)
- Added updateDoors export to be called with other teleporting commands to properly update doors when teleporting around (an issue since NUI doorlocks I've encountered again)
- Fixed a bug where when adding a door, every player thought they did it (so they did the animation). Now we pass the creatorSrc in 'qb-doorlock:client:newDoorAdded'
- Updated config.lua to have more of the possible configuration options listed
- Squashed bug where teleporting would not properly update nearby doors
- Added support for persistent door states (config option disabled by default), saved and loaded via json server side cross restart. Only door states different from the default will be saved. 
- Fixed some spelling errors

### CODE REVIEW

**Comment by BerkieBb** [comment]
Reviewing this tomorrow

**Review by iktray — APPROVED** [review]

**Comment by BerkieBb** [comment]
> After some mild testing, fixText being true and false both do the same thing for single doors? The code is the exact same as original nui_doorlocks so it may just be my doors being odd? It is detecting the variable properly but is never changing. Tested this with the DoorDebug drawtext showing at the location. Maybe someone else can test and verify for me. I need to look into it more.

fixText only really does something when the coords of the door is not in the middle, for example the bollards of gabz mrpd need that

**Comment by ItsANoBrainer** [comment]
I am aware of everything this does. I used nui_doorlock with ESX for about a year, and made all these customizations that I'm adding in this PR.

Changing fixText for me does not fix the text on doors where the text is not in the middle.

https://i.imgur.com/7YtH7h4.mp4

**Comment by BerkieBb** [comment]
After reviewing the code, the fixtext currently only works if you change objName to objHash because I forgot to add objName to that aswell for qb-doorlock v1 compatibility

**Comment by BerkieBb** [comment]
I have some adjustments for his pr, but it's too much to put in a review, so I'll pr to your repo

**Comment by milotiro95** [comment]
I just add your code into qb-doorlock and qb-input it seem to working well but there is no auto save into the config while creating a door ?

**Comment by ItsANoBrainer** [comment]
@milotiro95 What do you mean, can you elaborate what you mean?

**Comment by milotiro95** [comment]
Then I can use /createdoor aiming with my gun the door and it show the ID of the door, position, etc but there is no auto saving in the config.lua ou configs/example.lua 

Also while enable the selection of the door I need to restart to stop the door highlight.
If you need I can put picture to explain me better ^^".

**Comment by BerkieBb** [comment]
They do save after /newdoor, you're just doimg something wrong, go to the config and enable Config.Warnings then make a new door, after you've done that, don't restart the script or anything, just check the server console for messages

**Comment by milotiro95** [comment]
You're right I was doing wrong I just change the configs/example.lua for a new one like 1 and it's working. I was trying to put no value in job / gang / citizenID / object.
Which perm can I use for everyone can use the door ?

**Comment by BerkieBb** [comment]
Just add `allAuthorized = true,` to your door configuration

**Comment by milotiro95** [comment]
Thank you for the help BerkieBb it's work like a charm.

**Comment by PeeganKreece** [comment]
I downloaded the fork and have been fiddling with it. Only issue I’ve found is that check marking the door to be locked when you do /newdoor doesn’t actually apply it, you have to manually add it in the config. If I get to fixing this before you I’ll pull request it on your fork if that even works.

**Comment by BerkieBb** [comment]
> I downloaded the fork and have been fiddling with it. Only issue I’ve found is that check marking the door to be locked when you do /newdoor doesn’t actually apply it, you have to manually add it in the config. If I get to fixing this before you I’ll pull request it on your fork if that even works.

You need to update qb-input for that to work

**Comment by PeeganKreece** [comment]
> 

Will try, I’m glad it was user error.

**Comment by Andyauk** [comment]
with the new update I noticed an error from `Config.DoorStates = {}`

**Comment by milotiro95** [comment]
Same update not working well

**Comment by ItsANoBrainer** [comment]
Can you provide more useful information than “it doesn’t work?”. Screenshot the error would be helpful.

**Comment by milotiro95** [comment]
Due to the downtime on FiveM I cannot re do the update. I will give you the screenshot of the errors when FiveM work again.

**Comment by BerkieBb** [comment]
> Can you provide more useful information than “it doesn’t work?”. Screenshot the error would be helpful.

Go over my reviews and then you'll get it

**Comment by Andyauk** [comment]
> Reference in n



> Can you provide more useful information than “it doesn’t work?”. Screenshot the error would be helpful.

this is just putting the folder in from download.

```
[  script:qb-doorlock] Error parsing script @qb-doorlock/config.lua in resource qb-doorlock: @qb-doorlock/config.lua:52: '}' expected (to close '{' at line 51) near '='
[    c-scripting-core] Failed to load script config.lua.
[  script:qb-doorlock] Error loading script configs/example.lua in resource qb-doorlock: @qb-doorlock/configs/example.lua:2: attempt to index a nil value (global 'Config')
[  script:qb-doorlock] stack traceback:
[  script:qb-doorlock] 	@qb-doorlock/configs/example.lua:2: in main chunk
[    c-scripting-core] Failed to load script configs/example.lua.
[  script:qb-doorlock] Error loading script server/main.lua in resource qb-doorlock: @qb-doorlock/server/main.lua:330: attempt to index a nil value (global 'Config')
[  script:qb-doorlock] stack traceback:
[  script:qb-doorlock] 	@qb-doorlock/server/main.lua:330: in main chunk
[    c-scripting-core] Failed to load script server/main.lua.
[  script:qb-doorlock] SCRIPT ERROR: @qb-doorlock/server/main.lua:297: attempt to index a nil value (global 'Config')
[  script:qb-doorlock] > handler (@qb-doorlock/server/main.lua:297)
```

File: config.lua [hunk 1]
```diff
@@ -7,16 +7,49 @@ Config.Warnings = false -- if true it will show warnings in the console when cer
...
+	cantUnlock = true -- Set to true to not allow the player to unlock the door, only lock it. This means a script will have to trigger the unlock.
+	pickable = true, -- Can use a lockpick to unlock, only need if true
+	hideLabel = true, -- Set to true to hide the popup label, for hiding doors ;)
+    	remoteTrigger = true, -- If you want to be able to remote trigger a door with H, put this here 
+} 
+]]
+
 Config.DoorList = {
-	--- door1 for pacific opened with security card B
+Config.DoorStates = {}
```
Comment (FjamZoo) [inline config.lua:52]: The Config.DoorStates are within the Config.DoorList which breaks the entire Config for everyone.

**Comment by FjamZoo** [comment]
Fully tested, and works perfectly.

**Comment by Andyauk** [comment]
> Fully tested, and works perfectly.

Yes, great stuff

**Pull request merged**